feat: merge latest code to master

namnh-0652 opened the pull request:
- Merge latest code from Sun* DEV to master

namnh-0652 committed:
chore: do not pause when pip mode

namnh-0652 committed:
chore: move app to background when start PIP mode (like android)

namnh-0652 committed:
chore: invisible ios PIP placeholder

namnh-0652 committed:
chore: add check pip permission android

namnh-0652 committed:
fix: pause player when disable PIP permission

namnh-0652 committed:
Merge pull request #3 from thangtv-3206/android_pip_permission

[Add] Android pip permission

namnh-0652 committed:
chore: enable auto pip for android

namnh-0652 committed:
chore: enable auto pip ios

namnh-0652 committed:
fix: improve auto_pip in android

namnh-0652 committed:
chore: add events to handle pip status

namnh-0652 committed:
chore: improve pip mode

namnh-0652 committed:
chore: remove android pip status listener

thangtv-3206 committed:
Optimize toggle pip and rotation (#5)

* optimize animation on toggle PIP, detect rotation

namnh-0652 committed:
Merge branch 'develop' into auto_pip

namnh-0652 committed:
chore: update after merge code

tanth-1016 committed:
fix: Live Video freezes after repeatedly playing/pausing in PiP mode

namnh-0652 committed:
Merge pull request #7 from thangtv-3206/fix_livevideo_freeze_pip_ios

fix: Live Video freezes after repeatedly playing/pausing in PiP mode

Commit:
fix some bug rotate (#6)

namnh-0652 committed:
Merge branch 'develop' into auto_pip

thangtv-3206 committed:
Optimize PIP Android

thangtv-3206 committed:
fixup! Optimize PIP Android

thangtv-3206 committed:
update sample

thangtv-3206 committed:
Merge remote-tracking branch 'origin/thang/optimize_pip_android' into auto_pip

# Conflicts:
#	android/src/main/kotlin/uz/shs/better_player_plus/BetterPlayer.kt
#	android/src/main/kotlin/uz/shs/better_player_plus/BetterPlayerPlugin.kt
#	example/android/app/src/main/kotlin/com/jhomlala/better_player_example/MainActivity.kt

thangtv-3206 committed:
update after merged

Commit:
fix bug rotate in ios

namnh-0652 committed:
Merge pull request #10 from thangtv-3206/thang/merge_auto_pip

Merge auto pip

namnh-0652 committed:
chore: fix iOS multiple layers PIP

namnh-0652 committed:
chore: update sample

thangtv-3206 committed:
Merge pull request #8 from thangtv-3206/auto_pip

chore: auto Pip mode & some improvements

thangtv-3206 committed:
Merge pull request #11 from thangtv-3206/fix_bug_rotate_in_ios

fix bug rotate in ios

namnh-0652 committed:
fix: remove invalid constant

thangtv-3206 committed:
Merge pull request #12 from thangtv-3206/task/fix_error

fix: remove invalid constant

namnh-0652 committed:
chore: improve pip button iOS

thangtv-3206 committed:
Merge pull request #13 from thangtv-3206/improve_pip_button

chore: Improve pip button iOS

thangtv-3206 committed:
Optimize PIP (#16)

* optimize pip

thangtv-3206 committed:
Fix bug PIP (#18)

* Fix issue pip does not work at first time on IOS

* Auto PIP when app go to background on Android

thangtv-3206 committed:
Fix bug PIP IOS

thangtv-3206 committed:
Fix bug PIP Android

thangtv-3206 committed:
Format code

Commit:
fix bug can't update state video (#20)

thangtv-3206 committed:
Merge pull request #19 from thangtv-3206/thang/fix_bug_pip_ios

Fix bug PIP IOS

thangtv-3206 committed:
upgrade lib compatible Flutter 3.27.1

namnh-0652 committed:
chore: improve android handle audio focus

thangtv-3206 committed:
Merge pull request #21 from thangtv-3206/fix/handle_audio_focus

chore: improve android handle audio focus

thangtv-3206 committed:
Seek to latest when play

namnh-0652 committed:
chore: send pause event when lost audio focus

thangtv-3206 committed:
Merge pull request #23 from thangtv-3206/fix/handle_audio_focus

chore: [Android] send pause event when lost audio focus

thangtv-3206 committed:
Merge remote-tracking branch 'origin/develop' into seek_to_latest_when_play

thangtv-3206 committed:
fixup! Seek to latest when play

thangtv-3206 committed:
Merge pull request #22 from thangtv-3206/seek_to_latest_when_play

Seek to latest when play

thangtv-3206 committed:
feat: improve handling of automatic play/pause and PIP

- Modified automatic play/pause to consider only the app's lifecycle state (resumed) instead of checking for player visibility.
- Enhanced Picture-in-Picture (PIP) logic to start PIP when the video is either playing or buffering but not yet initialized, in addition to existing checks for visibility, PIP mode, and full-screen status.

thangtv-3206 committed:
feat: allow setting BetterPlayerAsmsTrack when setting up data source

- Added an optional `betterPlayerAsmsTrack` parameter to the `setupDataSource` method.
- Modified `setupDataSource` to use the provided `betterPlayerAsmsTrack` if available, otherwise using the default track.

thangtv-3206 committed:
feat: Improve ASMS data handling and live stream seeking

- Added handling for ASMS (Adaptive Streaming Media Services) tracks, subtitles, and audio tracks based on new flags (useAsmsTracks, useAsmsSubtitles, useAsmsAudioTracks).
- Modified iOS seeking logic for live streams to only seek to the live edge if it's a live stream and the difference between the current time and the live edge is greater than 1 second.

thangtv-3206 committed:
Upgrade libs

thangtv-3206 committed:
Optimize detect PIP Android

thangtv-3206 committed:
Upgrade libs Android

thangtv-3206 committed:
Setup dataSource with track

thangtv-3206 committed:
Merge pull request #24 from thangtv-3206/thang/optimize_pip

Optimize

thangtv-3206 committed:
feat: enable async ASMS track setup in BetterPlayer

- Modified `setupDataSource` to accept an optional `Future<BetterPlayerAsmsTrack?>?` named `betterPlayerAsmsTrackFuture`.
- Updated the ASMS track setup to await the `betterPlayerAsmsTrackFuture` before using it.
- Updated `picture_in_picture_page.dart` to demonstrate passing the ASMS track in setupDataSource in async mode.

thangtv-3206 committed:
feat: [iOS] Enable background audio playback for AVPlayer

- Set `audiovisualBackgroundPlaybackPolicy` to `AVPlayerAudiovisualBackgroundPlaybackPolicyContinuesIfPossible` to allow audio to continue playing when the app is in the background.

thangtv-3206 committed:
feat: Add Picture-in-Picture (PIP) support and improve device orientation handling

-   **Picture-in-Picture (PIP) Implementation:**
    -   Added PIP support for iOS and Android.
    -   Added `enablePIP` property to `BetterPlayerConfiguration` and pass to platform.
    -   Created a new `BetterPlayer` constructor in iOS with the ability to pass the `enablePIP` flag.
    - On Android, enable or not add `PlayerView` on `pipContainer` by `enablePIP` parameter.
-   - On Android, only enable pip mode when `playerView` not null.
-   - On Android, update `autoEnterEnabled` when player playing
    -   Updated `create` method in `MethodChannelVideoPlayer` and `VideoPlayerPlatformInterface` to accept `enablePIP` parameter.
    -   Updated `VideoPlayerController` constructor to accept `enablePIP` parameter.
-
-    **Device Orientation Handling:**
    -   Added `enterFullScreenWhenRotate` property to `BetterPlayerConfiguration`
    -   Removed `_deviceOrientationSubscription` when `enterFullScreenWhenRotate` is `false`.
-   - On Android, only seek to default position if is live stream.
-    **Other changes:**
-    - Example: add `enablePIP` in `BetterPlayerConfiguration`.
+   - Example: add `enablePIP` and `enterFullScreenWhenRotate` in `BetterPlayerConfiguration`.
+    - On Android, only seek to default position if it is a live stream.
+    - On Android, only enable pip mode when `playerView` is not null.
+    - On Android, update `autoEnterEnabled` when player is playing and `playerView` is not null.

thangtv-3206 committed:
fix: Prevent fullscreen toggle when in PIP mode

- Added a check to prevent entering fullscreen when device orientation changes while the player is in Picture-in-Picture (PIP) mode.
- Ensured that the fullscreen toggle logic is bypassed if `isPipMode()` returns true.

thangtv-3206 committed:
feat: Improve video player playback and position handling

- Modified the video player's periodic timer to stop updating if the player is disposed, buffering, or not playing.
- Used `Future.wait` to concurrently fetch `position` and `absolutePosition`.
- Added logic to early exit the periodic timer if the player is disposed, buffering or not playing
- Modified `seekTo` function to await the `play` and `pause` calls, allowing to stop update the timer when the video is not playing.

thangtv-3206 committed:
feat: Improve loading indicator behavior in Better Player

- Refactored loading detection logic by introducing an `isLoading()` method in `VideoPlayerController` to centralize and improve the logic.
- Updated `BetterPlayerMaterialControls` and `BetterPlayerCupertinoControls` to use the new `isLoading()` method for determining the loading state, instead of the previous custom logic.
- Modified the buffering logic in `VideoPlayerController` to consider the buffering interval before indicating that a video is loading.
- Updated `LiveVideoControls` to use `isLoading()` for more reliable loading state detection.
- Removed the custom `isLoading` method in `BetterPlayerControlsState` as it's no longer needed.
- Added the constant `_bufferingInterval` to `VideoPlayerController` to set the min. time of buffered video to hide loading timer.

thangtv-3206 committed:
fix: Update timer callback condition in video player

- Updated the timer callback condition to prevent execution when the video is not playing.
- Added a check for `!value.isPlaying` to avoid unnecessary operations when the video is paused.

thangtv-3206 committed:
fix: Improve seek logic to avoid immediate reset during buffering

- Modified the seek logic to avoid resetting the position during buffering.
- Added a condition to check if `value.isBuffering` is false before updating the seek position.

thangtv-3206 committed:
fix: Prevent rotation when player is not initialized or not visible

- Ensured that device rotation handling is skipped if the video is not initialized or if the player is not visible.

trungnt-2399 committed:
optimized scroll visibility

thangtv-3206 committed:
Merge pull request #25 from trungnt-2399/optimized_check_scroll_visibility

Optimized scroll visibility

thangtv-3206 committed:
feat: Enhance Picture-in-Picture functionality and lifecycle handling

trungnt-2399 committed:
Fix bug pau play pip Android

thangtv-3206 committed:
Merge pull request #26 from thangtv-3206/fix_bug_pause_play_pip

Fix bug pause play pip Android

thangtv-3206 committed:
fix: Refactor Picture-in-Picture lifecycle handling and initialization

thangtv-3206 committed:
fix: Refactor Picture-in-Picture handling and improve player visibility management

namnh-0652 commented on android/src/main/kotlin/uz/shs/better_player_plus/BetterPlayerPlugin.kt:
```diff
@@ -65,30 +57,48 @@ class BetterPlayerPlugin : FlutterPlugin, ActivityAware, MethodCallHandler,
+                                }
+                            )
+                        },
+                        ViewGroup.LayoutParams(
```
is it used?

thangtv-3206 replied:
em chủ đích dùng để cho cái View đấy hiển thị full màn hình luôn anh ạ

namnh-0652 commented on android/src/main/kotlin/uz/shs/better_player_plus/BetterPlayerPlugin.kt:
```diff
@@ -65,30 +57,48 @@ class BetterPlayerPlugin : FlutterPlugin, ActivityAware, MethodCallHandler,
                         currentBetterPlayer.disposeMediaSession()
-                        pipContainer.setPadding(0, 0, 0, 0)
-                        pipContainer.isVisible = false
+                        (pipContainer?.parent as ViewGroup)?.removeView(pipContainer)
```
`(pipContainer?.parent as? ConstraintLayout)?.removeView(pipContainer)`

thangtv-3206 replied:
chắc chắn `pipContainer?.parent` là `ViewGroup` - là class cha của `ConstraintLayout` và không thể null được a nhé

thangtv-3206 committed:
fix: Delay Picture-in-Picture activation to ensure proper state management

thangtv-3206 committed:
fix: Simplify Picture-in-Picture activation and improve state management

thangtv-3206 committed:
Optimize enter PIP iOS

thangtv-3206 committed:
Upgrade libs Android

thangtv-3206 committed:
feat: Add loading indicator and enhance PiP state management

This commit introduces a loading indicator for Picture-in-Picture (PiP) mode and improves overall PiP state handling.

Key changes:
- Added a loading indicator (ProgressBar) that displays in PiP mode when the video is not actively playing.
- Enhanced PiP state management by tracking the `isPip` status in `video_player.dart`.
- Improved Android PiP lifecycle handling, including safer view removal and listener management.
- Added logic to retry the data source if an error occurs while attempting to resume playback in PiP mode.
- Corrected MediaSession token handling on Android.

thangtv-3206 committed:
Refactor: Remove transition durations for full screen

This commit removes the explicit zero transition and reverse transition durations when pushing the full screen widget.

Key changes:
- Removed `transitionDuration: Duration.zero` from `PageRouteBuilder`.
- Removed `reverseTransitionDuration: Duration.zero` from `PageRouteBuilder`.

thangtv-3206 committed:
feat: Enable seamless resize for Picture-in-Picture on Android S and above

thangtv-3206 committed:
refactor: Replace custom visibility detector with built-in VisibilityDetector

thangtv-3206 committed:
refactor: Remove iOS-specific resetToOriginPipContentSource and enhance video player reload functionality

thangtv-3206 committed:
refactor: Simplify setBeforePipSourceRectHint and improve padding calculations for Picture-in-Picture

thangtv-3206 committed:
fixup! refactor: Remove iOS-specific resetToOriginPipContentSource and enhance video player reload functionality

hungnd-3027 committed:
Remove wakelock logic to let the app handle it itself

thangtv-3206 committed:
refactor: Set default volume to 0.0 in video player initialization

thangtv-3206 committed:
squash! Remove wakelock logic to let the app handle it itself

thangtv-3206 committed:
refactor: Remove gotoBackgroundWithPIP method and enhance Picture-in-Picture handling

hungnd-3027 committed:
Update logic rotate video

hungnd-3027 committed:
Update logic rotate video

thangtv-3206 committed:
feat: Add audio interrupt handling to resume playback on route change

thangtv-3206 committed:
feat: Enhance pause functionality to support promise playback resumption

thangtv-3206 committed:
fixup! feat: Enhance pause functionality to support promise playback resumption

thangtv-3206 committed:
fixup! feat: Enhance pause functionality to support promise playback resumption

Commit:
Ensure screen rotation syncs correctly with device orientation

Commit:
Update missing code

thangtv-3206 committed:
Merge pull request #27 from thangtv-3206/fix_issue_device_is_landscape_but_screen_is_rotate

Ensure screen rotation syncs correctly with device orientation

hungnd-3027 committed:
fix: Can not rotate after back in android (#28)

hungnd-3027 committed:
fix: orientation flicker after exiting fullscreen in landscape (#29)

Co-authored-by: hungnd-3027 <ngo.duc.hung@sun-asterisk,com>

thangtv-3206 committed:
fix: Prevent stalled check from executing when not playing and improve live playback seeking

thangtv-3206 committed:
fix: Correct playback rate handling to ensure accurate speed changes and prevent unintended playback state

thangtv-3206 committed:
chore: Update dependencies and SDK versions for improved compatibility

thangtv-3206 committed:
fix: Improve orientation handling to prevent unnecessary fullscreen transitions
